### PR TITLE
SCM inv source should trigger project update

### PR DESCRIPTION
**Pull request by fosterseth**
##### SUMMARY

scm based inventory sources should launch project updates prior to running inventory updates for that source (if update_on_launch is True for the project)

#### Implementation

The key change is to allow dependencies to have their own dependencies. A number of changes are needed to make that possible.

1. Turn off symmetry for `dependent_jobs` M2M field
It's better that this field capture some directionality of the relationship. This will help when we need to determine whether a job is blocked because a dependency has yet to run.

With change,
```python
job.dependent_jobs.add(proj)

job.dependent_jobs.all() # returns [proj]
proj.dependent_jobs.all() # returns []
proj.unifiedjobs_blocked_by.all() # returns [job]
```

2. Split up `generate_dependencies()` for Job and InventoryUpdate
This keeps the code cleaner and easier to read 

```python
if type(task) is Job:
    created_dependencies += self.gen_dep_for_job(task)
elif type(task) is InventoryUpdate:
    created_dependencies += self.gen_dep_for_inventory_update(task)
```

3. Call `generate_dependencies()` twice
Now that dependencies can have their own dependencies, we need to call generate_dependencies() a second time in the task manager loop. This need not be recursive, because the depth is only going to be a max of 2 (JobA > Inventory UpdateA > Project Update A)

Importantly, this all applies if you start an inventory source update manually. It doesn't have to be in the context of of a job launch. If a project `update_on_launch` is True, the project will be updated if corresponding job or inventory source is launched.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - API

##### AWX VERSION

```
awx: 20.0.2.dev249+g38ab7fc48c
```

### CODE REVIEW

File: awx/main/scheduler/dependency_graph.py
Comment (AlanCoding) [inline awx/main/scheduler/dependency_graph.py]: I'm not reading this coherently. Is `job` here an inventory update, or a job?

Also, the `InventoryUpdate` field that you want would be `source_project` IIRC, not `source`, so I'd expect to see reference to `.source_project_id`, but not totally sure what the thought here is.

File: awx/main/scheduler/dependency_graph.py
Comment (AlanCoding) [inline awx/main/scheduler/dependency_graph.py]: You should check if `job.inventory_source` is null before referencing `job.inventory_source.source_project_id`, because at this point it's not clear that it's SCM type, or we could be in a deletion cascade or something.

I'm also unsure if this condition is a good idea. Your other logic already covers creating and linking a project update dependency. This looks like it would block the inventory update if another update started after its dependency... which I don't really want.

File: awx/main/scheduler/task_manager.py
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py]: looks like you still need to change the method signature in some other places as well.

File: awx/main/scheduler/dependency_graph.py
Comment (fosterseth) [reply awx/main/scheduler/dependency_graph.py]: I mistakenly left this method in, the only method we need to update is `inventory_update_blocked_by`

File: awx/main/scheduler/task_manager.py
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py]: This could still need more work to avoid spawning the same update twice when the playbook comes from the same project. There's also an optimization issue here, because if the inventory update was just spawned, you would already have the update in memory, but this would re-fetch it.

Comment (AlanCoding) [reply awx/main/scheduler/task_manager.py]: but I'm not sure if the optimization part of this needs fixing.

File: awx/main/scheduler/dependency_graph.py
Comment (AlanCoding) [inline awx/main/scheduler/dependency_graph.py]: Still not sure if I agree with this. If the _generated update_ is still running, it will be blocked by

https://github.com/ansible/awx/blob/6b5aa1ec74630b27fd0125a4e28ce5c9f08530bb/awx/main/scheduler/task_manager.py#L109-L112

File: awx/main/scheduler/dependency_graph.py
Comment (fosterseth) [reply awx/main/scheduler/dependency_graph.py]: the task manager links the project update as a dependency to the _job_, not the inventory update. Maybe this should be a dependency off of the inventory update itself

**Comment by AlanCoding** [comment]
Thanks for the implementation notes. The CI check is telling you that you need to rebase and bump the migration number.

I'll try to get a good look at this tomorrow, and will want to sync up about the status of testing.

File: awx/main/models/unified_jobs.py
Comment (jbradberry) [inline awx/main/models/unified_jobs.py:579]: :fire:

File: awx/main/models/mixins.py
Comment (jbradberry) [inline awx/main/models/mixins.py]: This could be written more compactly

```
return not any(j.status in ['pending', 'waiting', 'running'] for j in self.dependent_jobs.all())
```

File: awx/main/scheduler/task_manager.py
Comment (fosterseth) [inline awx/main/scheduler/task_manager.py:305]: this was kind of misnamed -- our mixins have the method `get_jobs_fail_chain()` that is unique to types of tasks (ProjectUpdate, InventoryUpdate, Job, etc).

in essence all this method does is **add dependencies** with activity stream off, so that is what we should call it.

File: awx/main/scheduler/task_manager.py
Comment (jbradberry) [inline awx/main/scheduler/task_manager.py]: I'd turn this list comprehension into a generator expression (square brackets into parentheses)

File: awx/main/scheduler/task_manager.py
Comment (jbradberry) [inline awx/main/scheduler/task_manager.py]: How about just `if dependencies:`?

**Review by jbradberry — APPROVED** [review]

File: awx/main/models/mixins.py
Comment (AlanCoding) [reply awx/main/models/mixins.py]: Could also use the `ACTIVE_STATES` constant.

File: awx/main/scheduler/task_manager.py
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py:311]: Probably related to the migration, but could you explain how you're able to remove this?

Comment (fosterseth) [reply awx/main/scheduler/task_manager.py:311]: yeah this change here _could_ have implications

consider this setup (JobA and JobB both share the projUpdate dependency. But only JobA depends on invUpdate)

```
            JobA                    JobB
            |  |                      |
            |  |                      |
            |  |                      |
            |  |                      |
invUpdate <-+  +--->  projUpdate <----+
````

Currently on devel, invUpdate and projUpdate are linked (because of the above code snippet that is in question). If one fails, it brings down the other. If invUpdate fails, it causes projUpdate to fail, and thus JobB would fail, even though JobB never depended on invUpdate in the first place.

I can't think of why this behavior is needed.

Because of the nature of sharing dependencies, if a project update or inventory update fails, it should only fail the jobs that directly depend on it, not other stuff.

Comment (fosterseth) [reply awx/main/scheduler/task_manager.py:311]: the dependent_jobs field plays a role in determining which jobs fail, as I mention above.

The other role for dependent_jobs is to determine whether a job is blocked from running.

in task_manager.py
```python
if not task.dependent_jobs_finished():
    blocked_by = task.dependent_jobs.first()
```
On devel this really only comes into play for Jobs

type Job
```python
def dependent_jobs_finished(self):
    for j in self.dependent_jobs.all():
        if j.status in ['pending', 'waiting', 'running']:
            return False
    return True
```

But for ProjectUpdate and InventoryUpdate, this always returned True
```python
def dependent_jobs_finished(self):
    return True
```

Therefore, the removed code in the diff doesn't really have an effect here. Jobs still have the same dependent_jobs as before. The only difference is that InventoryUpdate and ProjectUpdate do not have each other as dependencies, but this doesn't affect the `is_blocked` behavior.

File: awx/main/models/mixins.py
Comment (AlanCoding) [inline awx/main/models/mixins.py:453]: I'm a little bit suspicious of this. I don't know that the `blocked_jobs` list is ordered... it might be, but that's really non-obvious looking at it, and the 0 reference, `blocked_jobs[0]` always feels suspicious. If this is well-established, then that's fine, but some code comments could help. What _is_ the first entry of the blocked jobs?

If the list is always length 1, then maybe you should just loop instead of using the first one?

Comment (AlanCoding) [reply awx/main/models/mixins.py:453]: oh that's probably the job. This is making more sense to me. It will fail its job if it fails... and any other unified jobs that is waiting on.

File: awx/main/scheduler/task_manager.py
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py]: Especially when looping over a global jobs list like `self.all_inventory_sources`, I would like to stick with `*_id` fields for performance reasons (even if it may or may not make a difference). So here, I would make the conditional be `if invsrc.inventory_id == task.inventory_id`

Comment (AlanCoding) [reply awx/main/scheduler/task_manager.py]: well I get that you're just moving the code

File: awx/main/scheduler/task_manager.py
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py:612]: I remember you mentioning this, but now I get it. Basically, `deps_of_deps` can only be the project updates generated for inventory updates. Because you know it only goes this-many levels deep it's written like this and not iterated indefinitely.

**Review by AlanCoding — CHANGES_REQUESTED** [review]
I have 2 requests remaining 
1. that we add some code comments to the effect that `blocked_jobs[0]` is just a back-reference to the job and
2. change related object references to the `*_id` field in a case, this will make future optimization work go more smoothly

after these, I'm happy to approve

File: awx/main/scheduler/task_manager.py
Comment (fosterseth) [reply awx/main/scheduler/task_manager.py:612]: exactly

File: awx/main/models/mixins.py
Comment (fosterseth) [reply awx/main/models/mixins.py:453]: yeah `blocked_jobs` are all the jobs that depend on this inventory update succeeding.

But, how can we fail the _other_ inventory updates that were started too (to retain the same behavior as devel)? Well, we can get that by looking at one of the blocked_jobs and getting its dependent_jobs. These dependent jobs might include project updates, so we need to filter it down further by checking its type

`if type(dep) is type(self) and dep.id != self.id:`

We can really grab _any_ of the blocked_jobs for this, as they all are guaranteed to have the same inventory, and thus the same set of inventory sources.

File: awx/main/scheduler/task_manager.py
Comment (fosterseth) [reply awx/main/scheduler/task_manager.py]: made the change

File: awx/main/models/mixins.py
Comment (fosterseth) [reply awx/main/models/mixins.py:453]: add some comments around this @AlanCoding

**Review by AlanCoding — APPROVED** [review]

File: awx/main/models/mixins.py
Comment (AlanCoding) [reply awx/main/models/mixins.py:453]: sounds good.

Ultimately, looks like these feed into `handle_work_error` via the dispatcher errback. That part of the design isn't bad, it's pretty sensible to handle the triggers this way. On the other hand, `handle_work_error` looks even worse than the reaper in terms for changing job status from underneath a running process, and not even bothering to send a cancel signal. Problem for another day...

**Comment by fosterseth** [comment]
my latest commit addresses an issue where the task manager can start a job even if a dependency has failed.
`job_blocked_by` checks if a dependency is in ACTIVE_STATES and if not, proceeds to run. Well that isn't good because the dependency could be a failed or error state.

The job would run and probably fail during execution because it detects that the last project update/inventory update has failed, stopping it in its tracks. But we can prevent the dispatcher from even getting that far by failing the task as soon as we know we should (in the task manager).

@AlanCoding please take a look at my most recent commit to see how I handled the above ^

https://github.com/ansible/awx/pull/12073/commits/b94b1ef430866a0764462067cee0b282210fa7d0

note: this is a current issue in devel as well

File: awx/main/scheduler/task_manager.py
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py:92]: My only note is that I'd like this message to be _distinguishable_ from the similar message in the handle_work_error task - for ease of debugging. Any changing of the wording whatsoever would be sufficient to accomplish that.

**Pull request merged**
